File: client/lib/core/purchases/purchases.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';

const questsAccessProductId = 'bogach.quests.access';
const multiplayer1ProductId = 'bogach.multiplayer.games.1';
const multiplayer10ProductId = 'bogach.multiplayer.games.10';
const multiplayer25ProductId = 'bogach.multiplayer.games.25';
const newYear2022ActionProductId = 'bogach.new_year_2022';

String hashProductId(String productId) {
  const salt = '(c) Bogach Team';
  final string = productId + salt;
  final bytes = utf8.encode(string);

  final hash1 = sha256.convert(bytes).toString();
  final hash2 = sha256.convert(utf8.encode(hash1)).toString();
  final hash3 = sha256.convert(utf8.encode(hash2)).toString();

  return hash3;
}

enum MultiplayerGamePurchases {
  oneGame,
  tenGames,
  twentyFiveGames,
}

extension MultiplayerGamePurchasesExtension on MultiplayerGamePurchases {
  String get productId {
    switch (this) {
      case MultiplayerGamePurchases.oneGame:
        return multiplayer1ProductId;

      case MultiplayerGamePurchases.tenGames:
        return multiplayer10ProductId;

      case MultiplayerGamePurchases.twentyFiveGames:
        return multiplayer25ProductId;

      default:
        return '';
    }
  }
}

File: server/functions/src/core/purchases/purchases.ts
// import { SHA256 } from 'crypto-ts';
import * as SHA256 from 'sha256';

export namespace Purchases {
  export const id = (productId: string) => {
    const salt = '(c) Bogach Team';
    const string = productId + salt;

    const hash1 = SHA256(string);
    const hash2 = SHA256(hash1);
    const hash3 = SHA256(hash2);

    return hash3;
  };

  export const questsAccessProductId = 'bogach.quests.access';

  export const newYear2022ActionProductId = 'bogach.new_year_2022';

  export const depricatedMultiplayer1ProductId = 'bogach.multiplayer.1';
  export const depricatedMultiplayer5ProductId = 'bogach.multiplayer.5';
  export const depricatedMultiplayer10ProductId = 'bogach.multiplayer.10';

  export const multiplayerGames1 = 'bogach.multiplayer.games.1';
  export const multiplayerGames10 = 'bogach.multiplayer.games.10';
  export const multiplayerGames25 = 'bogach.multiplayer.games.25';
}
